src/util/assignment.rs

```diff
@@ -0,0 +1,263 @@
+
+use crate::util::signature::timed_hmac_now;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Chunk {
+    pub id: String,
+    pub base_url: String,
+    pub files: HashMap<String, String>,
+    pub size_bytes: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Dataset {
+    pub id: String,
+    pub base_url: String,
+    pub chunks: Vec<Chunk>,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct Headers {
+    worker_id: String,
+    worker_signature: String,
+}
+
+#[serde_as]
+#[derive(Debug, Default, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+struct EncryptedHeaders {
+    #[serde_as(as = "Base64")]
+    identity: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    nonce: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    ciphertext: Vec<u8>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WorkerAssignment {
+    status: String,
+    chunks_deltas: Vec<u64>,
+    encrypted_headers: EncryptedHeaders,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Assignment {
+    datasets: Vec<Dataset>,
+    worker_assignments: HashMap<String, WorkerAssignment>,
+    #[serde(skip)]
+    chunk_map: Option<HashMap<String, u64>>,
+    #[serde(skip)]
+    pub id: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct NetworkAssignment {
+    pub(crate) url: String,
+    pub(crate) id: String,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct NetworkState {
+    pub(crate) network: String,
+    pub(crate) assignment: NetworkAssignment
+}
+
+impl Assignment {
+    pub fn add_chunk(&mut self, chunk: Chunk, dataset_id: String, dataset_url: String) {
+        match self.datasets.iter_mut().find(|dataset| dataset.id == dataset_id) {
+            Some(dataset) => dataset.chunks.push(chunk),
+            None => self.datasets.push(Dataset { id: dataset_id, base_url: dataset_url, chunks: vec![chunk] }),
+        }
+        self.chunk_map = None
+    }
+
+    pub async fn from_url(url: String, previous_id: Option<String>) -> Result<Option<Self>, Box<dyn std::error::Error>> {
+        let response_state = reqwest::get(url).await?;
+        let network_state: NetworkState = response_state.json().await?;
+        if Some(network_state.assignment.id.clone()) == previous_id {
+            return Ok(None);
+        }
+        let assignment_url = network_state.assignment.url;
+        let response_assignment = reqwest::get(assignment_url).await?;
+        let compressed_assignment = response_assignment.bytes().await?;
+        let mut decoder = GzDecoder::new(&compressed_assignment[..]);
+        let mut decompressed_assignment = String::new();
+        decoder.read_to_string(&mut decompressed_assignment)?;
+        let mut result: Assignment = serde_json::from_str(&decompressed_assignment)?;
```

Comment:
There is a `from_reader` function, no need for intermediate buffers
https://docs.rs/serde_json/latest/serde_json/fn.from_reader.html

Reply:
done